arangod/Metrics/ClusterMetricsFeature.cpp

```diff
@@ -68,73 +73,218 @@ void ClusterMetricsFeature::validateOptions(
           return;
         }
-        try {
-          update();
-        } catch (...) {
-          repeatUpdate();
-        }
-      });
+      } catch (...) {
+      }
+      [[fallthrough]];
+    case CollectMode::TriggerGlobal:
+      if (_count.fetch_add(kUpdate, std::memory_order_acq_rel) == 0) {
+        rescheduleUpdate(_timeout);
+      }
+    default:
+      break;
+  }
 }
 
 void ClusterMetricsFeature::update() {
-  auto& nf = server().getFeature<NetworkFeature>();
+  // We want more time than _timeout should expire
+  // after last try cross cluster communication
+  _lastUpdate = std::chrono::steady_clock::now();
+  uint64_t version = 0;
+  auto result = readData(version);
+  if (wasStop()) {
+    return;
+  }
+  _lastUpdate = std::chrono::steady_clock::now();
+  if (result == Result::Error) {
+    return rescheduleUpdate(std::max(_timeout, uint32_t{1}));
+  }
+  if (result == Result::New) {
+    return repeatUpdate();
+  }
   auto& cf = server().getFeature<ClusterFeature>();
-  // TODO try get from agency
+  auto& ci = cf.clusterInfo();
+  if (_acquiredUpdateLock) {
+    _acquiredUpdateLock = !ci.unlockMetricsUpdate();
+  }
+  if (_acquiredUpdateLock) {
+    return rescheduleUpdate(std::max(_timeout, uint32_t{1}));
+  }
+  _acquiredUpdateLock = ci.lockMetricsUpdate();
+  if (!_acquiredUpdateLock) {
+    return rescheduleUpdate(std::max(_timeout, uint32_t{1}));
+  }
+  if (wasStop()) {
+    return;
+  }
+  auto& nf = server().getFeature<NetworkFeature>();
   metricsOnCoordinator(nf, cf).thenFinal(
-      [this](futures::Try<RawDBServers>&& metrics) mutable {
-        // We want more time than kTimeout should expire
-        // after last try cross cluster communication
-        _lastUpdate = std::chrono::steady_clock::now();
-        if (!metrics.hasValue()) {
-          scheduleUpdate();
+      [this, version, &ci](futures::Try<RawDBServers>&& raw) mutable {
+        if (wasStop()) {
           return;
         }
+        auto result = Result::Old;
         try {
-          set(std::move(*metrics));
+          _lastUpdate = std::chrono::steady_clock::now();
+          result = writeData(version, std::move(raw));
+          _lastUpdate = std::chrono::steady_clock::now();
         } catch (...) {
+          result = Result::Error;
         }
-        repeatUpdate();
+        _acquiredUpdateLock = !ci.unlockMetricsUpdate();
+        _lastUpdate = std::chrono::steady_clock::now();
+        if (result != Result::Error) {
+          repeatUpdate();
+        }
+        rescheduleUpdate(std::max(_timeout, uint32_t{1}));
       });
 }
 
+ClusterMetricsFeature::Result ClusterMetricsFeature::readData(
+    uint64_t& version) {
+  auto vocbase = server().getFeature<SystemDatabaseFeature>().use();
+  aql::QueryString readText{
+      absl::StrCat("FOR m IN ", StaticStrings::MetricsCollection, " RETURN m")};
+  auto readQuery = aql::Query::create(
+      transaction::StandaloneContext::Create(*vocbase), readText, nullptr);
```

Comment:
we can use `std::shared_ptr<>` aliasing ctor to remove heap allocation:
```
transaction::StandaloneContext ctx{*vocbase};
  auto readQuery = aql::Query::create(
      std::shared_ptr<transaction::Context>{std::shared_ptr<transaction::Context>{}, &ctx}, readText, nullptr);
```

Reply:
Ok